client/lib/wp/sync-handler/cache-invalidation.js

```diff
@@ -0,0 +1,156 @@
+	 * @return {Promise} promise
+	 */
+	addItem( key, fn = () => {} ) {
+		return this.filterByKey( key, ( err, records = [] ) => {
+			debug( 'adding %o', key );
+
+			// add the fresh item into history list
+			records.unshift( { key, mark: new Date() } );
+			return localforage.setItem( RECORDS_LIST_KEY, records, fn );
+		},
+		err => {
+			fn( err );
+		} );
+	}
+
+	removeItem( key, fn = () => {} ) {
+		return this.filterByKey().then(
+			key, records => {
+				debug( 'adding %o', key );
+				return localforage.setItem( RECORDS_LIST_KEY, records, fn );
+			},
+			err => {
+				fn( err );
+			}
+		);
+	}
+
+	/**
+	 * retrieve all records filter by the given key
+	 *
+	 * @param {String} key - compare records with this key
+	 * @param {Function} [fn] - callback function
+	 * @return {Promise} promise
+	 */
+	filterByKey( key, fn = () => {} ) {
+		let changed = false;
+		return this.getAll().then( records => {
+			if ( ! records || ! records.length ) {
+				debug( 'No records stored' );
+				return fn( null, [] );
+			}
+
+			// filter records by the given key
+			records = records.filter( item => {
+				if ( item.key === key ) {
+					debug( '%o exists. Removing ...', key );
+					changed = true;
+				}
+				return item.key !== key;
+			} );
+
+			return fn( null, records, changed );
+		} );
+	}
+
+	/**
+	 * Calling this method all records will be removed.
+	 * It's a cleaning method and it should be used to re-sync the whole data.
+	 */
+	clean( fn = () => {} ) {
+		localforage.keys( ( err, keys ) => {
+			if ( err ) {
+				return fn( err );
+			}
+
+			keys.forEach( ( key ) => {
+				if ( ! ( /^sync-record-\w+$/.test( key ) ) ) {
+					return debug( '%o isn\'t a record', key );
+				}
+
+				localforage.removeItem( key ).then( () => {
+					debug( '%o has been removed', key );
+				}, removeErr => {
+					if ( removeErr ) {
+						return warn( removeErr );
+					}
+				} );
+			} );
+
+			localforage.removeItem( RECORDS_LIST_KEY ).then( () => {
+				debug( '%o has been removed', RECORDS_LIST_KEY );
+			}, removeListErr => {
+				if ( removeListErr ) {
+					return warn( removeListErr );
+				}
+			} );
+		} );
+	}
+
+	/**
+	 * Prune old records depending of the given lifetime
+	 *
+	 * @param {Number} lifetime - [records] lifetime ( in milliseconds )
+	 */
+	pruneRecordsFrom( lifetime = LIFETIME ) {
+		debug( 'start to prune records older than %sms', lifetime );
+		let updateRecords = false;
+
+		this.getAll()
+		.then( records => {
+			if ( ! records || ! records.length ) {
+				return debug( 'Records not found' );
+			}
+
+			records = records.filter( item => {
+				let reference = +new Date() - lifetime;
+				let timeago = moment( item.mark ).from();
+				if ( +item.mark < reference ) {
+					debug( '%o is too old (%s). Removing ...', item.key, timeago );
+					localforage.removeItem( item.key );
+					updateRecords = true;
+					return false;
+				}
+
+				return true;
+			} );
+
+			if ( updateRecords ) {
```

Comment:
Similar to my comment above, we can avoid the let here. We should be able to tell if something was modified if the length changed.

```javascript
const filteredRecords = records.filter( //... );
const updatedRecords = filteredRecords.length < records.length; 
```